### ISSUE TITLE
Menu bar is frozen when launching the app on macOS

### ISSUE THREAD

**Issue by matyalatte** [libui-ng/libui-ng#225]
Hi, I found an issue on macOS 10.15.

When I launch the example apps, the window will be focused but the menu bar will be frozen.
Nothing will happen when I put the mouse cursor on the apple icon.
So, to use the menu bar, I need to click another window or the desktop once and click the app window again.

I found some related posts on stackoverflow.
- [Cocoa application menu bar not clickable](https://stackoverflow.com/questions/33345686/cocoa-application-menu-bar-not-clickable)
- [Top Left Menu Bar Is Frozen when Main Window Shows Up](https://stackoverflow.com/questions/34469718/top-left-menu-bar-is-frozen-when-main-window-shows-up)

So, I tried to use `[uiprivNSApp() activateIgnoringOtherApps:YES];` on `uiInit` but it did't work.
I also tried to use it with the app delegate like this but it didn't work.
```c++
-(void)applicationWillFinishLaunching:(NSNotification *)aNotification
{
    [uiprivNSApp() setActivationPolicy:NSApplicationActivationPolicyRegular];
}

-(void)applicationDidFinishLaunching:(NSNotification *)notification
{
    [uiprivNSApp() activateIgnoringOtherApps:YES];
}
```

Anyone know how to solve this issue?

**Comment by cody271** [comment]
I can't reproduce this issue, menu correctly receives events on first try for me.

<img src="https://github.com/libui-ng/libui-ng/assets/11763424/b5be6077-ea51-4817-b772-5adc1bf6df86" width=323 height=498>

**Comment by matyalatte** [comment]
Hmm, the menu bar is wonking fine on my wxWigets app. So, I thought it's a libui issue...

Here is my environment.

- OS: 10.15.5
- Arch: x86_64
- Compiler: Apple clang 12.0.0
- Python: 3.10.6
- Meson: 1.2.1
- Meson Options: None (I mean, used default options)

And I made sure it can pass the unit tests.
I'm running mac on Windows with VirtualBox, so the VM might be the cause.
But I think there is a workaround because other GUI libraries are working fine.

**Comment by cody271** [comment]
> I'm running mac on Windows with VirtualBox, so the VM might be the cause.
>
Yes and that's an unsupported environment, so I can't really investigate any further..

> But I think there is a workaround because other GUI libraries are working fine.
>
If you can figure out a simple fix, we'll take another look at this issue.

**Comment by matyalatte** [comment]
I found another post about the menu bar. and it solved my problem.
[Why doesn't activate(ignoringOtherApps:) enable the menu bar?](https://stackoverflow.com/questions/62739862/why-doesnt-activateignoringotherapps-enable-the-menu-bar)

So, it's an issue of macOS 10.15.
and the policy setting should be defined in `applicationDidFinishLaunching`.

```objc
@implementation uiprivAppDelegate

...

-(void)applicationDidFinishLaunching:(NSNotification *)notification
{
    [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
}

@end
```

And I noticed wxWidgets is using the same method!
[wxWidgets/src/osx/cocoa/utils.mm at master · wxWidgets/wxWidgets](https://github.com/wxWidgets/wxWidgets/blob/master/src/osx/cocoa/utils.mm#L100)

**Comment by matyalatte** [comment]
I found out Eclipse has the same method.
[eclipse.platform.swt/bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java at master · eclipse-platform/eclipse.platform.swt](https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/bundles/org.eclipse.swt/Eclipse%20SWT/cocoa/org/eclipse/swt/widgets/Display.java#L5699)

I found it in tk as well.
[tk/macosx/tkMacOSXInit.c at main · tcltk/tk](https://github.com/tcltk/tk/blob/main/macosx/tkMacOSXInit.c#L209)

So, it should be applied to libui as well?
I'll send PR later.

**Comment by cody271** [comment]
I think I figured why I can't reproduce this issue: you're calling Cocoa APIs directly in your app.

Calling activate _before_ setting an activation policy is what triggers the freeze. If activate is never called, menu works normally.

Related PR that enables activation by default: https://github.com/libui-ng/libui-ng/pull/100

**Comment by cody271** [comment]
> And I noticed wxWidgets is using the same method!
> [wxWidgets/src/osx/cocoa/utils.mm at master · wxWidgets/wxWidgets](https://github.com/wxWidgets/wxWidgets/blob/master/src/osx/cocoa/utils.mm#L100)
>
The wxWidgets approach looks best to me: *none* of these hacks should be necessary if we detect that we're running in a proper `.app` bundle !

**Comment by matyalatte** [comment]
> Calling activate before setting an activation policy is what triggers the freeze. If activate is never called, menu works normally.

No, the master branch will make menu frozen, so `activateIgnoringOtherApps` was not the cause.
Just moving `setActivationPolicy` to `applicationDidFinishLaunching` resolved the issue in my case.

You can see my commit for the fix.
https://github.com/matyalatte/libui-ng/commit/e5923a7902f69d46cc0944a655a5b22fd947b9bc

As I posted before, eclipse and tk (and maybe wxWidgets as well?) had the same issues and they moved the policy setting to the app delegate to fix the issues.
So, I think it's a proper way to fix.

> The wxWidgets approach looks best to me

Yeah, so I should edit the app delegate like this?
```objc
-(void)applicationDidFinishLaunching:(NSNotification *)notification
{
	CFURLRef url = CFBundleCopyBundleURL(CFBundleGetMainBundle());
	CFStringRef path = CFURLCopyFileSystemPath (url, kCFURLPOSIXPathStyle);
	CFRelease(url);

	if (!CFStringHasSuffix(path, @".app")) {
		/* don't use setActivationPolicy outside this function
		 * or the menu bar won't work properly on macOS 10.15
		 * see https://github.com/libui-ng/libui-ng/issues/225
		 *
		 * don't check for a NO return;
		 * something (launch services?) causes running from application bundles to always return NO
		 * when asking to change activation policy, even if the change is to the same activation policy!
		 * see https://github.com/andlabs/ui/issues/6 */
		[uiprivNSApp() setActivationPolicy:NSApplicationActivationPolicyRegular];
	}
}
```

**Comment by cody271** [comment]
> Yeah, so I should edit the app delegate like this?

Yes, that looks mostly right. Few nits I see there:

`BOOL/YES/NO` not `bool/true/false` on Darwin.
 Missing `CFRelease(path);`, leaks?
Comment style is `//`

**Comment by cody271** [comment]
Also needs a comment to explain the edge case:

```objective-c
} else if ([uiprivNSApp() activationPolicy] == NSApplicationActivationPolicyAccessory) {
		// The app is an agent app that runs in the background and doesn't appear in the Dock.
		activate = true;
}
```

**Comment by matyalatte** [comment]
I sent the PR for the fix!

> Missing CFRelease(path);, leaks?

Idk how `CFStringRef` will work cause I just coiped the wxWidgets code.
But should I report it to the wxWidgets team?

**Comment by cody271** [comment]
> But should I report it to the wxWidgets team?

No I _think_ the wxWidgets code is safe. The smart pointer appears to release when it is out of scope:
```c++
wxString app = wxCFStringRef(path).AsString();
```

```c++
wxCFStringRef(WX_NSString ref)
        : wxCFRef< CFStringRef >((WX_OSX_BRIDGE_RETAINED CFStringRef) ref)
    {
    }
````

```c++
 ~wxCFRef()
    {   reset(); }

    void reset()
    {
        wxCFRelease(m_ptr);
        m_ptr = nullptr;
    }
```
